core/async/src/instrumentation/writer.rs

```diff
@@ -0,0 +1,157 @@
+use std::{
+    cell::{RefCell, RefMut},
+    collections::HashMap,
+    sync::Arc,
+    time::Instant,
+};
+
+use near_time::Clock;
+
+use crate::instrumentation::queue::InstrumentedQueue;
+use crate::instrumentation::{
+    WINDOW_SIZE_NS,
+    data::{ALL_ACTOR_INSTRUMENTATIONS, InstrumentedThread},
+};
+
+thread_local! {
+    static THREAD_WRITER: RefCell<Option<InstrumentedThreadWriter>> = const { RefCell::new(None) };
+}
+
+pub struct InstrumentedThreadWriter {
+    pub shared: Arc<InstrumentedThreadWriterSharedPart>,
+    current_window_start_time_ns: u64,
+    type_name_registry: HashMap<String, u32>,
+    target: Arc<InstrumentedThread>,
+}
+
+pub struct InstrumentedThreadWriterSharedPart {
+    actor_name: String,
+    clock: Clock,
+    reference_instant: Instant,
+    queue: Arc<InstrumentedQueue>,
+}
+
+impl InstrumentedThreadWriter {
+    pub fn start_event(&mut self, message_type: &str, dequeue_time_ns: u64) {
+        let start_time_ns =
+            self.shared.clock.now().duration_since(self.shared.reference_instant).as_nanos() as u64;
+        let message_type_id = if let Some(id) = self.type_name_registry.get(message_type) {
+            *id
+        } else {
+            let id = self.type_name_registry.len() as u32;
+            self.type_name_registry.insert(message_type.to_string(), id);
+            self.target.message_type_registry.push_type(message_type.to_string());
+            id
+        };
+        self.advance_window_if_needed_internal(start_time_ns);
+        self.target.start_event(message_type_id, start_time_ns, dequeue_time_ns);
+    }
+
+    // End an event for a specific message type.
+    // Note: easier to ask for message type here instead of storing a reverse mapping,
+    // as the caller knows it.
+    pub fn end_event(&mut self, _message_type: &str) {
+        let end_time_ns =
+            self.shared.clock.now().duration_since(self.shared.reference_instant).as_nanos() as u64;
+        self.advance_window_if_needed_internal(end_time_ns);
+        let _total_elapsed_ns = self.target.end_event(end_time_ns);
```

Comment:
note: `_message_type` and `_total_elapsed_ns` will later be used for metrics, but to keep the structure I kept them in this PR.